envoy/formatter/substitution_formatter.h

```diff
@@ -122,10 +123,13 @@ class CommandParserFactory : public Config::TypedFactory {
    * Creates a particular CommandParser implementation.
    *
    * @param config supplies the configuration for the command parser.
+   * @param context supplies the factory context.
    * @return CommandParserPtr the CommandParser which will be used in
    * SubstitutionFormatParser::parse() when evaluating an access log format string.
    */
-  virtual CommandParserPtr createCommandParserFromProto(const Protobuf::Message& config) PURE;
+  virtual CommandParserPtr
+  createCommandParserFromProto(const Protobuf::Message& config,
```

Comment:
A possibility to avoid the cascading change to all users of the API (including external extensions) - how about add a new class e.g. `CommandParserFactoryWithContext`, that's everything from this change, and make `CommandParserFactory` a subclass of `CommandParserFactoryWithContext`, that has a final override of the two-parameter version of the function that calls the single-parameter version of the function, and a PURE of that original function.

That way the existing API can be unchanged, but the new API is available for your needs here.

Reply:
I'd be open to that. Envoy has usually made internal breaking API changes with care since we'd still need to deprecate the non-context version at some point. My feeling (and verified with g3) was that these two extension points are rarely extended so it's easier to make the change now than go with graceful deprecation.

Thoughts? @yanavlasov or any other senior maintainer.

Reply:
If it's definitely required to deprecate it eventually then might as well rip the band-aid off now. But I was thinking it's not like you'd be maintaining two APIs - one would just be a convenience wrapper around the other, so that structure could easily be permanent.

(Similar to how http filters can be built on PassThruFilter for convenience, or on a parent with less defaults for precision. I don't think there's plans to deprecate one of those options just because there's redundancy.)

Reply:
Changing the factory base has implications on the registry since it's type-parametrized: see `REGISTER_FACTORY(CELFormatterFactory, CELFormatterFactory::CommandParserFactory)` and `Registry::InjectFactory<CommandParserFactory>`. I think we can do it but it might also expose some problems.

Reply:
Ah, that's a good point, so doing it that way would still require changes to all extensions. The only way to soften that blow then would be to make the existing base class have both functions with implementations that call the other, then whichever one you override would work (at the cost of it being a bit harder to understand what's going on, and no compile-time error if you forgot to override entirely).

Probably okay to just make it a quick pain for any existing extensions then, as implemented. (But that means it should have release notes for this reason.)

Reply:
@ravenblackx I'm not sure that the changelog is the right place. Those are oriented towards the end-users, and this change is purely internal, with no visible behavior, API, or bug fix. I deliberately tried to make it clear what to do when the sync failure happens for the projects using Envoy at HEAD.

Reply:
It's changing an extension-point's API. External users (such as me! We do!) can have an extension that uses this API.

Definitely in the `*Changes that may cause incompatibilities for some users, but should not for most*` section, but this change *would* break our build.

Reply:
@ravenblackx I understand this will cause problems for distributors of Envoy. But end-users are a different audience and are not affected by the internal API changes.

@yanavlasov Do you have some preference where to record the internal API changes? We can add a section to changelogs or a special field in the commit. Any thoughts?

Reply:
Not sure I agree with referring to this as "internal API changes". [REPO_LAYOUT.md](https://github.com/envoyproxy/envoy/blob/main/REPO_LAYOUT.md) says

> [envoy/](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/envoy/): "Public" interface headers for "core" Envoy.

i.e. `envoy/formatter/substitution_formatter.h` is a public interface, not an internal interface. (And it's this change that will break our build.)

We're not a distributor, we're an end-user that has custom extensions.

Reply:
I added a note under "area: build". That was used earlier and seems sufficient for this PR.

Reply:
Envoy has never promised to have stable code-level APIs, and we've never put code-level changes that don't change functionality into the changelog. FWIW, I maintain several out-of-tree extensions, and at least one of them won't compile with every version upgrade, and that's normal/expected. The intended audience of the changelog is people running Envoy, not people writing code in Envoy.